src/cmd/ksh93/sh/init.c

```diff
@@ -907,25 +907,50 @@ static_fn char *get_version(Namval_t *np, Namfun_t *fp) { return (nv_getv(np, fp
 // This is invoked when var `.sh.version` is used in a numeric context such as
 // `$(( .sh.version ))`.
 //
-// TODO: Figure out what we want this to do with a semantic version string and
-// a version string that may include information about the git commit used for
-// the build. The original version converted the YYYY-MM-DD date stamp into the
-// number represented by YYYYMMDD. But note it would return a nonsensical value
-// if a date stamp in the correct form did not start exactly 10 characters
-// before the end of the string referred to by`e_version`.
 static Sfdouble_t nget_version(Namval_t *np, Namfun_t *fp) {
     UNUSED(np);
-    const char *cp = SH_RELEASE;
-    int c;
+    char *cp = strdup(SH_RELEASE);
     Sflong_t t = 0;
+    char *dash;
+    char *dot;
+    char buff[256];
+    char *major_str, *minor_str, *patch_str;
+    int major, minor, patch;
     UNUSED(fp);
 
-    while ((c = *cp++)) {
-        if (c >= '0' && c <= '9') {
-            t *= 10;
-            t += c - '0';
-        }
-    }
+    // Version string in development version could be set like 2017.0.0-devel-1509-g95d59865
+    // If a '-' exists in version string, set it as end of string i.e. version string becomes
+    // 2017.0.0
+    dash = strchr(cp, '-');
+    if (dash) *dash = 0;
+
+    // Major version number starts at beginning of string
+    major_str = cp;
+
+    // Find the first '.' and set it to NULL, so major version string is set to 2017
+    dot = strchr(cp, '.');
+    *dot = 0;
+
+    // Minor version string starts after first '.'
+    minor_str = dot + 1;
+
+    // Find the second '.' and set it to NULL, so minor version string is set to 0
+    dot = strchr(minor_str, '.');
+    *dot = 0;
+
+    // Patch number starts after second '.'
+    patch_str = dot + 1;
+
+    major = atoi(major_str);
+    minor = atoi(minor_str);
+    patch = atoi(patch_str);
+
+    // Pad minor and patch numbers with 0's if they are less than 10
+    snprintf(buff, sizeof(buff), "%d%02d%02d", major, minor, patch);
```

Comment:
This approach is fine but did you consider just adding up the values? That is, `t = major * 10000 + minor * 100 + patch;`?

Reply:
You are right. We can get rid of `buff` array.